rfcs/0024-versioning-and-backwards-compatability.md

```diff
@@ -0,0 +1,349 @@
+
+### Versioned types and functors
+
+#### The problem
+
+Introducing versioned types inside functors removes a lot of the assurances we
+would like to make using the versioning system. Consider the following
+situation:
+
+```ocaml
+module A = struct
+  module Stable = struct
+    ...
+
+    module Latest = V1
+  end
+
+  type t = Stable.Latest.t
+end
+
+module Make (X : sig
+  type t [@@deriving bin_io, version]
+end) = struct
+  module B = struct
+    module Stable = struct
+      ...
+
+      module Latest = V1
+    end
+
+    type t = Stable.Latest.t
+  end
+end
+
+module V1_Make = Make (A.Stable.V1)
+```
+
+If the type `A.t` is moved to a new version, and then later the type `B.t` was
+moved to another version, then we would introduce
+
+```ocaml
+module V2_Make = Make (A.Stable.V2)
+```
+
+and have `V1_Make.B.V1.t`, `V1_Make.B.V2.t`, `V2_Make.B.V1.t`, and
+`V2_Make.B.V2.t`. This introduces the unused version `V1_Make.B.V2.t` and
+isolates the original version `V1_Make.B.V1.t` from the later versions.
+
+Similarly, if `B.t` was upgraded, followed by `A.t`, then the latest version
+`V2_Make.B.V2.t` is isolated from all earlier versions, and an unused version
+is introduced in `V2_Make.B.V1.t`.
+
+This also introduces a safety problem: it is possible to pass `A.Stable.Latest`
+to the functor. If `A.t` were upgraded in this situation, everything appears to
+be versioned, but we have unintentionally abandoned the old versions.
+
+#### The workaround
+
+Versioned types should not be used in the bodies of functors. Instead, the
+versioned types should be lifted out of the functor, and should accept type
+parameters for the versioned types that would have been introduced by the
+functor arguments. For example,
+
+```ocaml
+module Functor_type = struct
+  module Stable = struct
+    [%%versioned]
+
+    module V1 = struct
+      type 'a t = ...
+    end
+
+    module Latest = V1
+  end
+
+  type 'a t = 'a Stable.Latest.t
+end
+
+module Functor (A : sig
+  module Argument_type : sig
+    type t
+  end
+end) =
+struct
+  include Functor_type
+
+  type functor_type = Argument_type.t Functor_type.t
+
+  (* Code using `functor_type` *)
+
+  ...
+end
+```
+
+It is then easy to build a versioned type that combines the appropriate
+versions, without introducing any redundant types or lack of safety:
+
+```ocaml
+module Combined_type = struct
+  module Stable = struct
+    [%%versioned]
+
+    module V4 = struct
+      type t = Argument_type.Stable.V3.t Functor_type.Stable.V3.t
+    end
+
+    module Latest = V4
+
+    module V3 = struct
+      type t = Argument_type.Stable.V2.t Functor_type.Stable.V2.t
+
+      let to_latest t = ...
+    end
+
+    module V2 = struct
+      type t = Argument_type.Stable.V1.t Functor_type.Stable.V2.t
+
+      let to_latest t = ...
+    end
+
+    module V1 = struct
+      type t = Argument_type.Stable.V1.t Functor_type.Stable.V1.t
+
+      let to_latest t = ...
+    end
+  end
+
+  type t = Latest.t
```

Comment:
`Combined_type.t` will have `bin_io` (counter to our usual rule for types outside `Stable` modules ?

Reply:
I think we could give it `bin_io` implemented through `Module_version.Registration.Make.deserialize_binary_opt`, if you think that would be a useful thing to have?